nomad/leader.go

```diff
@@ -948,6 +959,88 @@ func (s *Server) restorePeriodicDispatcher() error {
 	return nil
 }
 
+// restoreLockTTLTimers iterates the stored variables and creates a lock TTL
+// timer for each variable lock. This is used during leadership establishment
+// to populate the in-memory timer.
+func (s *Server) restoreLockTTLTimers() error {
+
+	varIterator, err := s.fsm.State().Variables(nil)
+	if err != nil {
+		return fmt.Errorf("failed to list variables for lock TTL restore: %v", err)
+	}
+
+	// Iterate the variables, identifying each one that is associated to a lock
+	// and adding a TTL timer for each.
+	for varInterface := varIterator.Next(); varInterface != nil; varInterface = varIterator.Next() {
+		if realVar, ok := varInterface.(*structs.VariableEncrypted); ok && realVar.IsLock() {
+
+			// The variable will be modified in order to show that it no longer
+			// is held. We therefore need to ensure we perform modifications on
+			// a copy.
+			realVarCopy := realVar.Copy()
```

Comment:
I don't think I get this? `realVar` is itself a reference to `structs.VariableEncrypted`, is it not? By modifying it, won't we be modifying the underlying object?

Reply:
Yes, which is why we create a copy, so we are not modifying the underlying object which has been pulled from state and therefore is a pointer reference to the state object.

Reply:
right right 🤦